vignettes/target_types.Rmd

```diff
@@ -0,0 +1,49 @@
+---
+title: "Target type guidelines"
+author: "David Watkins"
+date: "3/25/2019"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Target types}
+  %\VignetteEngine{knitr::rmarkdown}
+  \usepackage[utf8]{inputenc}
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+```
+
+`scipiper` targets can be indicator files with a [shared cache](shared_cache.html), local data files with or without local indicator files, or simply R objects.  This vignette provides some guidelines for when to use these different patterns.
+
+## Objectives
+There are two main goals to keep in mind when deciding what style of target to use:
+ 
+ - **Minimize time spend unnecessarily rebuilding targets** --- i.e., those that have already been built locally or by collaborators
+ - **Keeping the pipeline simple** --- caching can save lots of build time, but also adds to the code that has to be maintained, can clutter the repo with build and indicator files, etc.
+ 
+ These two can sometimes be in tension, but usually it will be clear which is more important for a given case.
+ 
+## Key factors to consider
+
+ - **Time to create an object** vs **object size**: This is the biggest thing to consider --- is it faster to upload a file to the shared cache and have others download it, or just have others rebuild it themselves?  This will somewhat depend on your internet connection.  Don't forget about upload time as well.  Uploading is generally slower than downloading, especially for remote workers with residential internet connections.   
+ 
+ - **How often will a target be updated?**:  Saving a target to disk may be less useful if it is likely to need to be rebuilt before the file on disk is ever used again.
+ 
+ - **Collaborators**: Are all collaborators equipped to rebuild every step in the pipeline, if needed?  If a particular step needs to be done by certain people, say a processing-intensive munging step early in the pipeline, than this file will almost certainly need to be in the remote cache.
+ 
+## Target choices
+
+**Shared cache data/indicator file**: The target will be only be rebuilt if its dependencies have changed, and the indicator file has not been updated to trigger a download from the remote cache. For longer-running operations that are worthwhile to cache remotely.  See the [shared caching vignette](shared_cache.html) for different implementation options.
```

Comment:
Clarification: even if the indicator file has been updated, the data file target will only be rebuilt (i.e., downloaded) if the data file is explicitly required and requested for some other operation. This is because .yml files designed for a shared cache [should] only tell remake about dependencies among .ind files, and they don't/shouldn't tell remake about how data files relate to one another. Which means that if I only need some data file far downstream of some indicator file, I should never actually need to download that file from the remote cache at all. In those cases, the uploading and storage are issues but the downloading cost is probably pretty low for the team as a whole.

Agreed on the primary use case (use for longer-running operations that are worthwhile to cache remotely).

However, I think there's also a secondary use case for having an indicator file. If you don't use an indicator file, then everybody needs to build the data file (or object) not just to be able to use that file/object directly, but also to be able to acquire or build _any_ file _anywhere_ downstream of the data file/object. So git-committed indicator files can be useful when you're coding a whole bunch of data targets that nobody else is going to need to use, and all they want is the output of your chain of analyses. Even if each of those data targets is relatively quick to build on its own, the whole set of targets can take a while. In that situation, using indicator files would save everybody else from having to build all of the intermediate data files, allowing them to jump straight to the files they actually need.

For the secondary use case, I suspect there's a way to include indicator files without actually pushing the data files to a shared cache, and I think we've maybe sorta done this with task plan outputs in the past, but the right method is still pretty fuzzy to me; it'd be great to get a recommendation written out.